### PR TITLE
Fix issue #441

### PR DESCRIPTION
**Pull request by K1li4nL**
Attempt to fix issue #441 by adding the original seed to the XOF struct and defining Reset()

### CODE REVIEW

File: xof/blake2xb/blake.go [hunk 1]
```diff
@@ -33,7 +34,11 @@ func New(seed []byte) kyber.XOF {
 			panic("blake2b.XOF.Write should not return error: " + err.Error())
 		}
 	}
-	return &xof{impl: b}
+	seedCopy := make([]byte, len(seed))
+	copy(seedCopy, seed)
+	x := xof{impl: b, seed: seedCopy}
+
+	return &x
```
Comment (AnomalRoil) [inline xof/blake2xb/blake.go:41]:
```suggestion
	return &xof{impl: b, seed: seedCopy}
```

File: xof/blake2xb/blake.go [hunk 2]
```diff
@@ -61,6 +66,11 @@ func (x *xof) Reseed() {
 	x.impl = y.(*xof).impl
 }
 
+func (x *xof) Reset() {
+	y := New(x.seed)
+	x.impl = y.(*xof).impl
```
Comment (AnomalRoil) [inline xof/blake2xb/blake.go:71]: This is making too many useless allocations.
Please, refactor `New` to use `Reset` after having done the `blake2b.NewXOF` step,  and use the fact that `impl` has its own `Reset` function if you can.

File: xof/blake2xs/blake.go [hunk 3]
```diff
@@ -61,6 +66,11 @@ func (x *xof) Reseed() {
 	x.impl = y.(*xof).impl
 }
 
+func (x *xof) Reset() {
+	y := New(x.seed)
```
Comment (AnomalRoil) [inline xof/blake2xs/blake.go:70]: Same as above. Use Reset in New, not the other way around.

File: xof/keccak/keccak.go [hunk 4]
```diff
@@ -36,6 +39,11 @@ func (x *xof) Reseed() {
 	x.sh.Write(x.key)
 }
 
+func (x *xof) Reset() {
+	x.sh.Reset()
+	x.sh.Write(x.seed)
```
Comment (AnomalRoil) [inline xof/keccak/keccak.go:44]: yes, this is nice

**Review by AnomalRoil — CHANGES_REQUESTED** [review]

File: xof/blake2xb/blake.go (hunk 2 above)
Comment (K1li4nL) [reply xof/blake2xb/blake.go]: Should be fixed now !

File: xof/blake2xs/blake.go (hunk 3 above)
Comment (K1li4nL) [reply xof/blake2xs/blake.go]: Should be fixed now !

File: xof/xof_test.go [hunk 5]
```diff
@@ -214,6 +215,43 @@ func testReseed(t *testing.T, s kyber.XOFFactory) {
 	}
 }
 
+func TestResetNoSeed(t *testing.T) {
+	for _, impl := range impls {
```
Comment (AnomalRoil) [inline xof/xof_test.go:219]: I think this is only testing `blake2xb`, no? Could you add `blake2xs` to the list of `impls` too?

**Review by AnomalRoil — CHANGES_REQUESTED** [review]

File: xof/blake2xs/blake.go [hunk 6]
```diff
@@ -61,6 +65,13 @@ func (x *xof) Reseed() {
 	x.impl = y.(*xof).impl
 }
 
+func (x *xof) Reset() {
+	x.impl.Reset()
+	if len(x.seed) > blake2s.Size {
+		x.impl.Write(x.seed[blake2s.Size:])
+	}
+}
```
Comment (AnomalRoil) [inline xof/blake2xs/blake.go:73]: Is it really necessary to store the entire seed then? Wouldn't storing just `seed2` be enough and test here if its len is bigger than 0?

**Review by AnomalRoil — CHANGES_REQUESTED** [review]
Okay, one last comment I guess

File: xof/blake2xs/blake.go (hunk 6 above)
Comment (K1li4nL) [reply xof/blake2xs/blake.go]: Now that I am not using New() in Reset() as you suggested, you are right we can do that. Fixed !

**Review by AnomalRoil — APPROVED** [review]
lgtm

**Pull request merged**